### PR TITLE
2524 implement rate limiter 

### PR DESCRIPTION
**Pull request by cyberliem**

### CODE REVIEW

File: lib/etherscan/etherscan_client_test.go [hunk 1]
```diff
@@ -37,3 +40,39 @@ func TestEtherScanClient(t *testing.T) {
 	assert.NoError(t, err, fmt.Sprintf("get normal tx error: %v", err))
 	assert.NotEmpty(t, normalTxs, "get normal txs failed")
 }
+
+func TestEtherScanClientWithRateLimiter(t *testing.T) {
+	//commented out this and run with different rps to see its behaviour
```
Comment (favadi) [inline lib/etherscan/etherscan_client_test.go]: if this test need to be observed by human eyes, it should be always disabled and only manually enable by developer when need. Please replace testutil.SkipExternal with t.Skip().

File: lib/etherscan/etherscan_client_test.go [hunk 2]
```diff
@@ -37,3 +40,39 @@ func TestEtherScanClient(t *testing.T) {
...
+	etherscanClient := etherscan.New(etherscan.Network("api"), etherscanAPIKey)
+	limiter := rate.NewLimiter(rate.Limit(float64(rps)), 1)
+	etherscanClient.BeforeRequest = limitRate(limiter)
+
+	for i := 0; i < 20; i++ {
+		wg.Add(1)
+		go func(index int) {
+			defer wg.Done()
+			_, err := etherscanClient.EtherTotalSupply()
+			sugar.Debugw("finshied a request", "request_number", index, "finish_time", time.Now(), "error", err)
```
Comment (favadi) [inline lib/etherscan/etherscan_client_test.go]: finshied, typo.

File: lib/binance/binance_client.go [hunk 3]
```diff
@@ -24,17 +26,19 @@ const (
...
-	sugar     *zap.SugaredLogger
+	APIKey      string
+	SecretKey   string
+	sugar       *zap.SugaredLogger
+	rateLimiter *rate.Limiter
 }
 
 //NewBinance return a new client for binance api
-func NewBinance(apiKey, secretKey string, sugar *zap.SugaredLogger) *Client {
+func NewBinance(apiKey, secretKey string, sugar *zap.SugaredLogger, limiter *rate.Limiter) *Client {
```
Comment (favadi) [inline lib/binance/binance_client.go]: We should have an interface for limiter, maybe

```go
type binanceLimiter interface {
  WaitN(n uint64)
}
```
And the limiter should be a an optional configuration with default request per second taken from `https://api.binance.com/api/v1/exchangeInfo`.

File: lib/huobi/huobi_client.go [hunk 4]
```diff
@@ -25,17 +27,19 @@ const (
...
-	sugar     *zap.SugaredLogger
+	APIKey      string
+	SecretKey   string
+	sugar       *zap.SugaredLogger
+	rateLimiter *rate.Limiter
 }
 
 //NewClient return a new HuobiClient instance
-func NewClient(apiKey, secretKey string, sugar *zap.SugaredLogger) *Client {
+func NewClient(apiKey, secretKey string, sugar *zap.SugaredLogger, limiter *rate.Limiter) *Client {
```
Comment (favadi) [inline lib/huobi/huobi_client.go]: same comment as binance client.

Comment (cyberliem) [reply lib/huobi/huobi_client.go]: changed to interface

File: lib/binance/binance_client.go (hunk 3 above)
Comment (cyberliem) [reply lib/binance/binance_client.go]: changed to . interface

**Pull request merged**